server/src/internalClusterTest/java/org/elasticsearch/search/SearchWithIndexBlocksIT.java

```diff
@@ -0,0 +1,129 @@
+        for (String index : unblockedIndices) {
+            expectedHits += numDocsPerIndex.get(index);
+        }
+
+        BytesReference pitId = null;
+        try {
+            OpenPointInTimeRequest openPITRequest = new OpenPointInTimeRequest(indices.toArray(new String[0])).keepAlive(
+                TimeValue.timeValueSeconds(10)
+            ).allowPartialSearchResults(true);
+            pitId = client().execute(TransportOpenPointInTimeAction.TYPE, openPITRequest).actionGet().getPointInTimeId();
+            SearchRequest searchRequest = new SearchRequest().source(
+                new SearchSourceBuilder().pointInTimeBuilder(new PointInTimeBuilder(pitId).setKeepAlive(TimeValue.timeValueSeconds(10)))
+            );
+            assertHitCount(client().search(searchRequest), expectedHits);
+        } finally {
+            if (pitId != null) {
+                client().execute(TransportClosePointInTimeAction.TYPE, new ClosePointInTimeRequest(pitId)).actionGet();
+            }
+        }
+    }
+
+    private List<String> createIndices() {
+        int numIndices = randomIntBetween(1, 3);
+        List<String> indices = new ArrayList<>();
+        for (int i = 0; i < numIndices; i++) {
+            indices.add("test" + i);
+            createIndex("test" + i);
+        }
+        return indices;
+    }
+
+    private Map<String, Integer> indexDocuments(List<String> indices) {
+        Map<String, Integer> numDocsPerIndex = new HashMap<>();
+        List<IndexRequestBuilder> indexRequests = new ArrayList<>();
+        for (String index : indices) {
+            int numDocs = randomIntBetween(0, 10);
+            numDocsPerIndex.put(index, numDocs);
+            for (int i = 0; i < numDocs; i++) {
+                indexRequests.add(prepareIndex(index).setId(String.valueOf(i)).setSource("field", "value"));
+            }
+        }
+        indexRandom(true, indexRequests);
+
+        return numDocsPerIndex;
+    }
+
+    private List<String> addIndexRefreshBlockToSomeIndices(List<String> indices) {
+        List<String> unblockedIndices = new ArrayList<>();
+        var blocksBuilder = ClusterBlocks.builder().blocks(EMPTY_CLUSTER_BLOCK);
+        for (String index : indices) {
+            boolean blockIndex = randomBoolean();
+            if (blockIndex) {
+                blocksBuilder.addIndexBlock(ProjectId.DEFAULT, index, IndexMetadata.INDEX_REFRESH_BLOCK);
+            } else {
+                unblockedIndices.add(index);
+            }
+        }
+
+        var dataNodes = clusterService().state().getNodes().getAllNodes();
+        for (DiscoveryNode dataNode : dataNodes) {
+            ClusterService clusterService = internalCluster().getInstance(ClusterService.class, dataNode.getName());
+            ClusterState currentState = clusterService.state();
+            ClusterState newState = ClusterState.builder(currentState).blocks(blocksBuilder).build();
+            setState(clusterService, newState);
```

Comment:
This method is not intended to be used in integration test as it overrides the current data node cluster state.

For testing the INDEX_REFRESH_BLOCK I think it makes sense to only have unit tests in stateful elasticsearch.